assembly/src/assembler/mod.rs

```diff
@@ -140,7 +140,33 @@ impl Assembler {
         let module = module.compile_with_options(options)?;
         assert_eq!(module.kind(), kind, "expected module kind to match compilation options");
 
-        self.module_graph.add_module(module)?;
+        self.module_graph.add_ast_module(module)?;
+
+        Ok(())
+    }
+
+    /// Adds the compiled library to provide modules for the compilation.
+    pub fn add_compiled_library(&mut self, library: CompiledLibrary) -> Result<(), Report> {
+        let module_indexes: Vec<ModuleIndex> = library
+            .into_module_infos()
+            .map(|module| self.module_graph.add_module_info(module))
+            .collect::<Result<_, _>>()?;
+
+        self.module_graph.recompute()?;
+
+        // Register all procedures as roots
+        for module_index in module_indexes {
+            for (proc_index, proc) in
+                self.module_graph[module_index].unwrap_info().clone().procedure_infos()
+            {
+                let gid = GlobalProcedureIndex {
+                    module: module_index,
+                    index: proc_index,
+                };
+
+                self.module_graph.register_mast_root(gid, proc.digest)?;
+            }
+        }
```

Comment:
A couple of thought here (maybe neither is for this PR):

1. It may be cleaner if most of this is inside the `ModuleGraph`. Basically, we could have something like `ModuleGraph::add_compiled_modules(module_infos)` and updating of internal data structures could be encapsulated inside the `ModuleGraph`.
2. It seems like we are paying in complexity here for things that are probably won't be used. Specifically, the typical scenario is that we'd instantiate the assembler with some compiled libraries and then start assembling things. Here, we allow for a scenario where we add compiled libraries in the middle of assembly process. So, maybe a better approach would be to specify the set of compiled libraries in the assembler constructor and then the `ModuleGraph` itself would be just instantiated with compiled modules. But this may be a bigger change to the API.

Reply:
For example, we could have the following constructors:
```rust
pub fn new(kernel: Kernel, libraries: IntoIterator<Item = CompiledLibrary>) -> Self {
    todo!("implement")
}

pub fn with_kernel(kernel: Kernel) -> Self {
    todo!("implement")
}

pub fn with_libraries(libraries: IntoIterator<Item = CompiledLibrary>) -> Self {
    todo!("implement")
}
```
And `Assembler::default()` could still be used to instantiate an empty assembler (i.e., the assembler without kernel or libraries).

Reply:
I applied the 1st suggestion but created #1416 for the 2nd suggestion (which would be better to do as we remove the old `Library` struct or after)